Action Manager | Registration API Setup

AMZN-daimini opened the pull request:
Introduces the core APIs to register Action Contexts and Actions with the Action Manager.

This is a barebones implementation - some parameters (like parent identifier for contexts, or iconPath for actions) are part of some of the function signatures but are not functional. I'm trying to keep the PRs relatively small and incremental to simplify reviews :)

Note that this change won't affect any of the Editor behavior right now, since the changes are gated behind a setreg flag that defaults to false (O3DE/ActionManager/EnableNewActionManager).

As part of the development, I also locally introduced a barebones Menu Manager and verified that the current implementation works, and it is possible to add the same action to multiple menus as theorized during planning. (Menu manager changes are coming separately)
![HabemusBasicFunctionality](https://user-images.githubusercontent.com/82231674/166343783-28284446-3ef9-4e46-93d0-a09595373284.gif)

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Action/ActionManager.cpp:
```diff
@@ -0,0 +1,93 @@
+        if (!widget)
+        {
+            return AZ::Failure(
+                AZStd::string::format("Action Manager - Could not register action context \"%s\" to a null widget.", identifier.data())
```
You can't do this sorry 😮 `.data()` might not be null terminated, so this could go horribly wrong. There is a way to do this for format strings - `"%.*s", stringLength, pointerToString`, but be super careful just calling `.data()` without a size (I'll send you an article to it offline 😝)

cgalvan commented on Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Action/ActionManager.cpp:
```diff
@@ -0,0 +1,93 @@
+            ));
+        }
+
+        m_actions.insert(
```
Should we have a warning if you try to register an `identifier` that's already registered?

cgalvan commented on Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Action/ActionManagerInterface.h:
```diff
@@ -0,0 +1,52 @@
+        ) = 0;
+
+        //! Retrieve a QAction via its identifier.
+        virtual QAction* GetAction(AZStd::string_view actionIdentifier) = 0;
```
Since this is a getter we should make this `const`

hultonha replied:
Don't do this unless it also returns a `const` object (otherwise it's considered 'const broken') - https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/const.md#classes-of-const-in_correctness

cgalvan commented on Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Action/EditorActionContext.h:
```diff
@@ -0,0 +1,37 @@
+        AZStd::string m_parentIdentifier;
+        AZStd::string m_name;
+
+        QWidget* m_widget;
```
Since this has a default constructor as well, we should initialize this to `nullptr`

cgalvan commented on Code/Framework/AzToolsFramework/Tests/ActionManager/ActionManagerTests.cpp:
```diff
@@ -0,0 +1,93 @@
+        action->trigger();
+        EXPECT_TRUE(actionTriggered);
+    }
+
```
If we do want to do a warning about registering an `identifier` that already exists (see above comment), we should add a unit test for that case as well.

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Action/EditorActionContext.cpp:
```diff
@@ -0,0 +1,27 @@
+namespace AzToolsFramework
+{
+    EditorActionContext::EditorActionContext(
+        AZStd::string_view identifier, AZStd::string_view name, AZStd::string_view parentIdentifier, QWidget* widget)
```
suggestion: As we're making copies of the strings here, we might want to actually just pass `AZStd::string` here by value and then `AZStd::move` the values into the member variables? Not a huge deal but could be clearer

hultonha commented on Code/Framework/AzToolsFramework/Tests/ActionManager/ActionManagerFixture.cpp:
```diff
@@ -0,0 +1,30 @@
+
+namespace UnitTest
+{
+    void ActionManagerFixture::SetUpEditorFixtureImpl()
```
Awesome to be adding tests at this point 😄 Great to see 🙂

hultonha reviewed (commented):
Overall looking like a great start, just a few small gripes that should be quick to fix 👍

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Action/ActionManager.cpp:
```diff
@@ -0,0 +1,93 @@
+
+        m_actionContexts.insert(
+            {
+                identifier.data(),
```
Sounds like a plan, thanks!

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Action/ActionManager.h:
```diff
@@ -0,0 +1,49 @@
+    {
+    public:
+        ActionManager();
+        virtual ~ActionManager();
```
Ah perfect thanks! 👍

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Action/EditorAction.cpp:
```diff
@@ -0,0 +1,40 @@
+        AZStd::string_view name,
+        AZStd::string_view description,
+        AZStd::string_view category,
+        AZStd::function<void()> handler)
```
Oh yes you're quite right 😅 my bad, thanks!

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Action/ActionManager.cpp:
```diff
@@ -0,0 +1,109 @@
+        return nullptr;
+    }
+
+    const QAction* ActionManager::GetActionConst(AZStd::string_view actionIdentifier)
```
nit: This function can be annotated `const` as well

`const QAction* ActionManager::GetActionConst(AZStd::string_view actionIdentifier) const`

AMZN-daimini replied:
I have actually tried this, but it doesn't build - operator [] on the map doesn't like it. I'll keep this in mind and try it out in follow-up PRs if it makes sense.

hultonha commented on Code/Framework/AzToolsFramework/AzToolsFramework/ActionManager/Action/ActionManagerInterface.h:
```diff
@@ -0,0 +1,53 @@
+
+    //! ActionManagerInterface
+    //! Interface to register and trigger actions in the Editor.
+    class ActionManagerInterface
```
suggestion: As you want to inherit from this, you actually might want to add the virtual destructor here (sorry I missed that before), if you don't want to delete the object from a base pointer you can make the destructor protected and non-virtual (which is the preferred approach for EBuses)

AMZN-daimini replied:
I have seen that approach for EBuses, but not for Interfaces - is it actually used somewhere?
I'm likely going to move on with this submission as is, but I can definitely fix this up in a follow-up.

hultonha reviewed (approved):
Couple of very minor suggestions, otherwise looks like a great start! 😄

AMZN-daimini commented:
I was having trouble with string_view to string conversions. So I reconsidered the needs of the functions in the API:
- Register functions have checks and error messages that work better with const string& (string format doesn't like %.s, and operator[] on the map has the same issue). If they succeed, are going to create an EditorAction or EditorActionContext object - both of which have string properties. In those cases, we'll have to create a copy whether the arguments are const string& or string_view.
- Get functions use operator[] on the map, so same issue as above.
In all cases, using const string& results in less string instantiations. I also introduced Tom's suggestion of creating a copy of the string in the constructors for EditorAction and EditorActionContext, and then moving them.

With all this considered, I have changed the APIs to use const string& instead of string_view, and I feel much better about the implementations since I don't need weird conversions or new instantiations everywhere :) But let me know if I missed anything.

hultonha commented on Code/Framework/AzToolsFramework/Tests/ActionManager/ActionManagerFixture.cpp:
```diff
@@ -0,0 +1,33 @@
+
+    void ActionManagerFixture::TearDown()
+    {
+
+
```
uber nit: Remove this line break?

hultonha reviewed (approved):
:+1:

hultonha commented:
> I was having trouble with string_view to string conversions. So I reconsidered the needs of the functions in the API:
> 
>     * Register functions have checks and error messages that work better with const string& (string format doesn't like %.s, and operator[] on the map has the same issue). If they succeed, are going to create an EditorAction or EditorActionContext object - both of which have string properties. In those cases, we'll have to create a copy whether the arguments are const string& or string_view.
> 
>     * Get functions use operator[] on the map, so same issue as above.
> 
> 
> In all cases, using const string& results in less string instantiations. I also introduced Tom's suggestion of creating a copy of the string in the constructors for EditorAction and EditorActionContext, and then moving them.
> 
> With all this considered, I have changed the APIs to use const string& instead of string_view, and I feel much better about the implementations since I don't need weird conversions or new instantiations everywhere :) But let me know if I missed anything.

If you're interesting in avoiding temporaries for look-ups in the map (probably a premature optimization at this point but you never know) I'd definitely recommend looking up Heterogeneous Lookup in Associative Containers (see this link https://abseil.io/tips/144 and https://www.cppstories.com/2021/heterogeneous-access-cpp20/ for more details). It's pretty cool and can mean you avoid constructing a `std::string` from a `const char*` for example when doing a key look-up.

AMZN-daimini commented:
> > I was having trouble with string_view to string conversions. So I reconsidered the needs of the functions in the API:
> > ```
> > * Register functions have checks and error messages that work better with const string& (string format doesn't like %.s, and operator[] on the map has the same issue). If they succeed, are going to create an EditorAction or EditorActionContext object - both of which have string properties. In those cases, we'll have to create a copy whether the arguments are const string& or string_view.
> > 
> > * Get functions use operator[] on the map, so same issue as above.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In all cases, using const string& results in less string instantiations. I also introduced Tom's suggestion of creating a copy of the string in the constructors for EditorAction and EditorActionContext, and then moving them.
> > With all this considered, I have changed the APIs to use const string& instead of string_view, and I feel much better about the implementations since I don't need weird conversions or new instantiations everywhere :) But let me know if I missed anything.
> 
> If you're interesting in avoiding temporaries for look-ups in the map (probably a premature optimization at this point but you never know) I'd definitely recommend looking up Heterogeneous Lookup in Associative Containers (see this link https://abseil.io/tips/144 and https://www.cppstories.com/2021/heterogeneous-access-cpp20/ for more details). It's pretty cool and can mean you avoid constructing a `std::string` from a `const char*` for example when doing a key look-up.

Thanks! I  think the current changes remove that issue for now, but this is definitely interesting and could turn out useful :D
